component/async/component.go

```diff
@@ -184,15 +206,28 @@ func (c *Component) processing(ctx context.Context) error {
 	}
 }
 
+func (c *Component) dispatchMessage(msg Message) error {
+	if c.concurrency > 1 {
+		c.jobs <- msg
+		return nil // a message failure is safe to ignore in parallel message consumption
+	}
+	return c.processMessage(msg)
+}
+
 func (c *Component) processMessage(msg Message) error {
 	err := c.proc(msg)
 	if err != nil {
 		return c.executeFailureStrategy(msg, err)
 	}
-
 	return msg.Ack()
 }
 
+func (c *Component) worker(messages <-chan Message) {
+	for msg := range messages {
+		_ = c.processMessage(msg)
```

Comment:
Not handling an error is obviously not a good idea. The component as it was did stop processing and would fallback to a component retry loop. This change breaks this functionality.

Reply:
The error is handled. The processMessage function on the component will execute the failure strategy and even print logs. In case the failure strategy also returns an error, meaning that we failed to ack/nack the message, it doesn't matter because out-of-order queues have dead letter boxes and visibility timeouts. Async out-of-order message consumers are meant to go into retry loops, it's not like consuming an event stream in-order. I hate to say this, but "it's a feature". 😄 

So as long as we change the code to disallow concurrency for in-order queues, this worker implementation is not problematic. I hope that is clear or do you need further clarifications?

Reply:
The previous implementation returned an error from the process message which in turn returned the error from the component which then stopped patron.
Now the parallel processing will always return nil which means that the component will never return an error because the error will be ignored. as you can see the  processMessage return error from the failure strategy and also if the ACK of the message does not succeed...
I believe we should handle this in a more correct way because failure strategey will never terminate the component and message ACK is not handled at all...

Reply:
Message ACK is handled. It's only a failure of the ACK/NACK that would be ignored, which means the message will reach the visibility timeout and end up in the queue's retry mechanism, then into dead letter after N retries. Which is expected behaviour. So why is this code not correct?

The component will still stop after X retries if it gets an HTTP error when retrieving messages from the queue.

Reply:
I understand this is confusing, which is why having completely separate async components for in-order and out-of-order brokers makes sense. I spent some time explaining this to my own team members before it was clear, do you want to discuss this over Zoom?

Reply:
In any case we should add a comment in the code explaining that for future readers, because I had the same question on our Zoom calls, and future readers probably will, too.